bin/src/main.rs

```diff
@@ -177,29 +169,48 @@ async fn main() {
         tokio::spawn(offset_state.run().unwrap());
     }
 
-    let fs_source = fs_source
-        .process(&mut fs_tailer_buf)
-        .expect("except Failed to create FS Tailer")
-        .filter_map(|r| async {
-            match r {
-                Err(e) => {
-                    match e {
-                        fs::cache::Error::WatchOverflow => {
-                            error!("{}", e);
-                            panic!("overflowed kernel queue");
-                        }
-                        fs::cache::Error::PathNotValid(path) => {
-                            debug!("Path is not longer valid: {:?}", path);
-                        }
-                        _ => {
-                            warn!("Processing inotify event resulted in error: {}", e);
-                        }
-                    };
-                    None
-                }
-                Ok(lazy_lin_ser) => Some(StrictOrLazyLineBuilder::Lazy(lazy_lin_ser)),
+    let ds_source_params = (
+        config.log.dirs.clone(),
+        config.log.rules.clone(),
+        config.log.lookback.clone(),
+        initial_offsets.clone(),
+    );
+
+    let fs_source = tail::RestartingTailer::new(
+        ds_source_params,
+        |item| match item {
+            Err(fs::cache::Error::WatchOverflow) => {
+                warn!("overflowed kernel queue, restarting stream");
+                return true;
             }
-        });
+            _ => false,
+        },
+        |params| {
+            let watched_dirs = params.0.clone();
```

Comment:
This is the core of the change. It creates new copies of the params that are then owned by the tailer. The `process` method now actually consumes the tailer but generates a stream of events.